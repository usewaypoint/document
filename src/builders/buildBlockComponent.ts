import React from 'react';
import { z } from 'zod';

import { BaseZodDictionary, DocumentBlocksDictionary } from '../utils';

export default function buildBlockComponent<T extends BaseZodDictionary>(blocks: DocumentBlocksDictionary<T>) {
  type BaseBlockComponentProps<TType extends keyof T> = {
    type: TType;
    data: z.infer<T[TType]>;
  };

  return function BlockComponent({ type, data }: BaseBlockComponentProps<keyof T>): React.ReactNode {
    return React.createElement(blocks[type].Component, data);
  };
}
