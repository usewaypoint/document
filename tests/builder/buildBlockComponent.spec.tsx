import React from 'react';
import { z } from 'zod';

import { render } from '@testing-library/react';

import buildBlockComponent from '../../src/builders/buildBlockComponent';

describe('builders/buildBlockComponent', () => {
  it('renders the specified component', () => {
    const BlockComponent = buildBlockComponent({
      SampleBlock: {
        schema: z.object({ text: z.string() }),
        Component: ({ text }) => <div>{text.toUpperCase()}</div>,
      },
    });
    expect(render(<BlockComponent type="SampleBlock" data={{ text: 'Test text!' }} />).asFragment()).toMatchSnapshot();
  });
});
