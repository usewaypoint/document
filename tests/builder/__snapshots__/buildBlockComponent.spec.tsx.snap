// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`builders/buildBlockComponent renders the specified component 1`] = `
<DocumentFragment>
  <div>
    TEST TEXT!
  </div>
</DocumentFragment>
`;
